Global Audit View: Vulnerabilities

rbt-mm opened the pull request:
### Description

This PR introduces introduces new API endpoints in the `FindingResource` of the backend.  These two new endpoints filter every finding by ACLs and other optional filters and then returns them , either by occurrence of vulnerability or grouped by vulnerability.

Those endpoints are used in a new view, which is introduced in the [Frontend PR](https://github.com/DependencyTrack/frontend/pull/411), and grant a simple way of gathering all findings in one place and filtering/sorting them by different criteria. 

### Addressed Issue

#1770

<!-- REQUIRED
    Reference the issue addressed by this PR, e.g. `#1234`.
    Use keywords like `closes` or `fixes` to signal that this PR resolves the issue,
    causing the issue to be closed automatically when the PR is merged:
        https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->

### Additional Details

- Requires the `VIEW_VULNERABILITY` permission
- Tested on internal H2 Database, MSSQL and PostgreSQL

![localhost_8081_vulnerabilityAudit (3)](https://user-images.githubusercontent.com/113189967/217294540-2b00fce9-c57e-4f0f-bd5e-ff5a298243df.png)

![localhost_8081_vulnerabilityAudit (4)](https://user-images.githubusercontent.com/113189967/217294646-593169f8-77e9-4b72-ab8c-70e4e06ff866.png)


**A PR for a policy violations audit will soon follow!**

<!-- OPTIONAL
    If desired, share more technical details about the change here.
    Elaborating on why you implemented the change the way you did can be super helpful to the reviewer.
    Did you consider other solutions? Any problems you ran into along the way?
-->

### Checklist

<!-- REQUIRED
    Mark items in this list as done by adding a `x` between the square brackets.
    Non-applicable items may be marked as such by surrounding their text with tildes (`~`).

    This is not meant to be a strict to-do list. If you're unsure about anything,
    just leave it empty for now. The maintainers are happy to assist you in figuring it out!
-->

- [x] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/github-templates/CONTRIBUTING.md#pull-requests)
~- [ ] This PR fixes a defect, and I have provided tests to verify that the fix is effective~
- [x] This PR implements an enhancement, and I have provided tests to verify that it works as intended
~- [ ] This PR introduces changes to the database model, and I have added corresponding [update logic](https://github.com/DependencyTrack/dependency-track/tree/master/src/main/java/org/dependencytrack/upgrade)~
~- [ ] This PR introduces new or alters existing behavior, and I have updated the [documentation](https://github.com/DependencyTrack/dependency-track/tree/master/docs/_docs) accordingly~

walterdeboer commented on src/main/java/org/dependencytrack/resources/v1/FindingResource.java:
```diff
@@ -172,5 +175,115 @@ public Response analyzeProject(
+            filters.put("cvssTo", cvssTo);
+            final List<Finding> findings = qm.getAllFindings(filters, showSuppressed, showInactive);
+            return Response.ok(findings).header(TOTAL_COUNT_HEADER, findings.size()).build();
+        }
```
@sonatype-lift ignore

walterdeboer commented on src/main/java/org/dependencytrack/resources/v1/FindingResource.java:
```diff
@@ -172,5 +175,115 @@ public Response analyzeProject(
+            filters.put("aggregatedAttributedOnDateTo", aggregatedAttributedOnDateTo);
+            final List<GroupedFinding> findings = qm.getAllFindingsGroupedByVulnerability(filters, showInactive);
+            return Response.ok(findings).header(TOTAL_COUNT_HEADER, findings.size()).build();
+        }
```
@sonatype-lift ignore

valentijnscholten commented:
I tried it, but whatever I put in the filters on the left, or Clear All, I always get _No Matching records found_.

rkg-mm commented:
> I tried it, but whatever I put in the filters on the left, or Clear All, I always get _No Matching records found_.

Thats strange since I tested it too.. you have projects with vulnerabilities and your account also has permissions to audit vulnerabilities?

valentijnscholten commented:
I am logged in as `admin` with all permissions assigned. And I am seeing the menu entry in the left menu bar.

My main usecase for this view is to see the most recently found vulnerabilities and the projects that are affected. So even without permissions to audit vulnerabilities, it would be useful to be able to use this view.

rkg-mm commented:
> I am logged in as `admin` with all permissions assigned. And I am seeing the menu entry in the left menu bar.
> 
> My main usecase for this view is to see the most recently found vulnerabilities and the projects that are affected. So even without permissions to audit vulnerabilities, it would be useful to be able to use this view.

Ah yes, my mistake, as Richard mentioned in the PR description: VIEW_VULNERABILITY is enough and should fit your use case.

> EDIT: Hmm all projects have disappeared, let me check what's going on.
That sounds like the reason then

valentijnscholten commented:
Yeah, projects are back and I can see vulnerabilities. But the initial page load takes a very long time. The download size is over 14MB for that xhr call.

rkg-mm commented:
Damn. That sounds a bit too much. We don't by chance have some huge example BOMs prepared that could be shared for contributors to test changes on larger databases? Our test setup is too small it seems :D

valentijnscholten commented:
I have ~200 projects and ~6000 vulnerabilities. Might be that that big join without pagina is slow, or that DataNucleus performs a couple of extra queries on each finding to get missing data. 

Here's a link a 22MB BOM with 9k components: https://github.com/DependencyTrack/dependency-track/issues/1905#issuecomment-1224764675

rkg-mm commented:
Hat big JOIN might take some time - but the result is still only 25 (or whatever you configured) entries, which should even for 100 rows not be 14MB

melba-lopez commented:
@rkg-mm are you able to fix/address the sonatype bugs that were identified?

rkg-mm commented:
> @rkg-mm are you able to fix/address the sonatype bugs that were identified?

@rbt-mm will provide a reworked version soon, that fixes the performance issues and will also address sonatype issues. Please don't merge for now.

rbt-mm commented:
This PR is now updated to include the following changes:
- Changed pagination from client side to server side, resulting in faster load times
- Changed hierarchical ACL processing to the default ACL processing that is currently used in Dependency-Track (database specific code is now removed)

Please check this updated feature out again and leave your feedback!

nscuro commented on src/main/java/org/dependencytrack/persistence/FindingsSearchQueryManager.java:
```diff
@@ -0,0 +1,374 @@
+                case "attributedOnDateFrom" -> processRangeFilter(queryFilter, params, filter, filters.get(filter), "\"FINDINGATTRIBUTION\".\"ATTRIBUTED_ON\"", true, true, false);
+                case "attributedOnDateTo" -> processRangeFilter(queryFilter, params, filter, filters.get(filter), "\"FINDINGATTRIBUTION\".\"ATTRIBUTED_ON\"", false, true, false);
+                case "textSearchField" -> processInputFilter(queryFilter, params, filter, filters.get(filter), filters.get("textSearchInput"));
+                case "cvssFrom" -> processRangeFilter(queryFilter, params, filter, filters.get(filter), "\"VULNERABILITY\".\"CVSSV3BASESCORE\"", true, false, false);
+                case "cvssTo" -> processRangeFilter(queryFilter, params,filter, filters.get(filter), "\"VULNERABILITY\".\"CVSSV3BASESCORE\"", false, false, false);
```
We'll need equivalent filters for CVSSv2, and in the near future also CVSSv4. Naming the filter (and input fields in the frontend) `cvssFrom` and `cvssTo` could be a bit confusing.

nscuro reviewed (commented):
Thanks for the PR, and addressing all comments so far @rbt-mm and @rkg-mm!

A few additional comments from my side regarding the implementation.

As for the feature itself, it certainly makes sense to me. However, especially for reporting, requirements and desired features can vary quite a bit between organizations. Before we ship this feature, I'd really love more community input. I posted a request for it in our Slack channel: https://owasp.slack.com/archives/C6R3R32H4/p1692563653439049

We discussed this feature in our last DT maintainers meeting, and we decided to *not* include it in the upcoming 4.9 release. For that, we're mostly waiting for a few small blockers to unblock (e.g. https://github.com/DependencyTrack/dependency-track/issues/2850), but otherwise we want to get that release shipped ASAP.

As mentioned earlier, we believe that this feature needs more eyes on it, and also more testing. The aim would be to get it merged as soon as 4.9 is out of the door. We're aware of the resource constraints on your side, but given the current state and quality of the code in this PR (props to @rbt-mm!), we're optimistic that we can finish the work ourselves if required.

rkg-mm commented:
> Thanks for the PR, and addressing all comments so far @rbt-mm and @rkg-mm!
> 
> A few additional comments from my side regarding the implementation.
> 
> As for the feature itself, it certainly makes sense to me. However, especially for reporting, requirements and desired features can vary quite a bit between organizations. Before we ship this feature, I'd really love more community input. I posted a request for it in our Slack channel: https://owasp.slack.com/archives/C6R3R32H4/p1692563653439049
> 
> We discussed this feature in our last DT maintainers meeting, and we decided to _not_ include it in the upcoming 4.9 release. For that, we're mostly waiting for a few small blockers to unblock (e.g. #2850), but otherwise we want to get that release shipped ASAP.
> 
> As mentioned earlier, we believe that this feature needs more eyes on it, and also more testing. The aim would be to get it merged as soon as 4.9 is out of the door. We're aware of the resource constraints on your side, but given the current state and quality of the code in this PR (props to @rbt-mm!), we're optimistic that we can finish the work ourselves if required.

Thanks for the Feedback, OK let's wait for community input here

rkg-mm commented on src/main/java/org/dependencytrack/resources/v1/FindingResource.java:
```diff
@@ -177,5 +181,115 @@ public Response analyzeProject(
+                                   @ApiParam(value = "Show suppressed findings")
+                                   @QueryParam("showSuppressed") boolean showSuppressed,
+                                   @ApiParam(value = "Filter by severity")
+                                   @QueryParam("severity") String severity,
```
I would prefer to not omit it, since it's an important filter that I will need. When I first testet this feature we did a small calculation how much we would miss, and it was a very small percentage that is affected by #2474. Maybe a small hint in the UI, that there are exceptional cases due to technical reasons where this filter will provide incorrect results until a future version could solve this?

rkg-mm commented on src/main/java/org/dependencytrack/model/GroupedFinding.java:
```diff
@@ -0,0 +1,99 @@
+                "VULNERABILITY"."PUBLISHED",
+                "VULNERABILITY"."CWES",
+                COUNT(DISTINCT "PROJECT"."ID") AS "AFFECTED_PROJECT_COUNT",
+                MIN("AFFECTEDVERSIONATTRIBUTION"."FIRST_SEEN") AS "FIRST_OCCURRENCE",
+                MAX("AFFECTEDVERSIONATTRIBUTION"."LAST_SEEN") AS "LAST_OCCURRENCE"
```
ok in that case let's remove this

nscuro commented:
For folks willing to test this, I pushed container images to Docker Hub:

* `dependencytrack/apiserver:1770-global-audit-view`
* `dependencytrack/frontend:1770-global-audit-view`

melba-lopez commented:
thanks for this feature @rbt-mm and @rkg-mm !!  I 💟 where this is going and would see this being extremely useful! Most of my questions below i made as a "code" formatting for ease of finding.  

I see the business need especially for PSIRT and CISO organizations.  One thing that dawned on me as I was experimenting with the new dashboard is _I wish i had a "roll up" view_.  For smaller organizations or maybe even a few projects, I feel the way it is could suffice.  However, as the number of projects get bigger or number of users/personas get bigger, we would want some sort of summary.  Then being able to click on that summary to drill down into the areas of interest. I'm not sure how this would be presented to the user visually or the type of data.  However, I did not see an item specifically pertaining to this, so not sure if this should be in a new PR or not.  

My DT is still syncing with nvd etc, so i think it may still be processing in the background to populate the dashboard/metrics.  If after its done, I still see broken summaries we may want to revisit the code to see where it might need to be tweaked.  

![image](https://github.com/DependencyTrack/dependency-track/assets/101211710/83af8400-eda6-44bf-93ae-636246f46c56)
![image](https://github.com/DependencyTrack/dependency-track/assets/101211710/8b340026-92e2-4882-9f4b-5451b03d410b)
![image](https://github.com/DependencyTrack/dependency-track/assets/101211710/c42628a2-1f73-48ea-9953-ed2e516b6fcf)
![image](https://github.com/DependencyTrack/dependency-track/assets/101211710/b1ab2f21-cd73-4ec5-9f14-3ea044953f33)
![image](https://github.com/DependencyTrack/dependency-track/assets/101211710/3aaf15df-98fb-47a8-b642-a625d9b4a5dc)
![image](https://github.com/DependencyTrack/dependency-track/assets/101211710/5a8ddeec-7275-4ef0-8f18-9e5c8308244e)

When i go to the new dashboard, i'm a bit confused by the occurances vs groups.  `Could we add a mini description in the ui (maybe a hover) to explain to the user the purpose of these two views?`  `Instead of 2 views, should it be an option for the user?`  Also, doing a spot check with moment, I see 8 distinct packages with 8 critical vulnerabilities (see screenshot above), but in the occurances dashboard i only see 2.  Similarly the same output happens with "grouped" dashboard.  I filtered on high/critical so was surprised `i didn't get a minimum of 8 critical (in at least one of the dashboards).  Should this be expected?`

![image](https://github.com/DependencyTrack/dependency-track/assets/101211710/a19eb351-4630-4474-8a3f-cca2a5eb2173)
![image](https://github.com/DependencyTrack/dependency-track/assets/101211710/94fab51b-c1e8-4472-abc7-8adaddd845e5)

Thanks @nscuro for publishing a snapshot of this feature!  Was fun playing with it and excited for its future release 😄

rkg-mm commented:
@melba-lopez I think the missing vulnerabilties are due to the issue nscuro pointed out already, basically this: https://github.com/DependencyTrack/dependency-track/issues/2474 we probably need to change that behaviour in parallel to make the filtering work correct, but that seems to be a smaller change to me.
Can you check if you see all of them without severity filter?

rkg-mm commented:
@nscuro apart from an additional change for #2474, anything else you would need for this PR? I'll try to get someone to work on #2474 next weeks.

rkg-mm commented:
We just created a PR for #2474 -> https://github.com/DependencyTrack/dependency-track/pull/3151
After this PR I think this feature should be mergeable :)

nscuro commented:
Thanks @rkg-mm, I will have a look shortly.

rkg-mm commented:
@nscuro Can you add this to 4.11 milestone?

nscuro commented:
@rkg-mm / @rbt-mm Could you rebase this PR on top of the current `master` please, to resolve the merge conflict? Also as of #3408, a solution for the last blocker for this PR has been merged.

rkg-mm commented:
@nscuro should be fine now

nscuro reviewed (approved):
Looks and behaves good. 

Thanks @rkg-mm and @rbt-mm, and deep apologies for taking this long to get this over the line. Really appreciate your contributions.
